src/Chainweb/Sync/WebBlockHeaderStore.hs

```diff
@@ -232,24 +241,48 @@ getBlockHeaderInternal headerStore payloadStore candidateHeaderCas candidatePayl
                     return $! (Nothing, x)
                 (Just !x) -> return $! (maybeOrigin, x)
 
+        -- Check that the chain id is correct. The candidate cas is indexed just
+        -- by the block hash. So, if this fails it is most likely a bug in code
+        -- that uses or populates the candidateHeaderCas.
+        --
+        unless (_chainId header == cid) $ throwM $ GetBlockHeaderFailure
+            $ "chain id of block header doesn't match expected chain id. "
+            <> "Most likely, this is a bug in Chainweb.Sync.WebBlockHeaderStore."
+
+        -- Perform intrinsic validations on the block header. There's another
+        -- complete pass of block header validations after payload validation
+        -- when the header is finally added to the db.
+        --
+        now <- getCurrentTimeIntegral
+        validateIntrinsicM now header
+
         -- Query Prerequesits recursively. If there is already job for this
         -- prerequesite in the memo-table it is awaited, otherwise a new job is
         -- created.
         --
-        let queryPrerequesiteHeader p = Concurrently $ void $ do
-                logg Debug $ taskMsg k $ "getBlockHeaderInternal.getPrerequisteHeader for " <> sshow h <> ": " <> sshow p
+        let queryAdjacentParent p = Concurrently $ void $ do
+                logg Debug $ taskMsg k $ "getBlockHeaderInternal.getPrerequisteHeader (adjacent) for " <> sshow h <> ": " <> sshow p
                 getBlockHeaderInternal headerStore payloadStore candidateHeaderCas candidatePayloadCas priority maybeOrigin' p
 
+            -- Perform inductive (involving the parent) validations on the block
+            -- header. There's another complete pass of block header validations
+            -- after payload validation when the header is finally added to the db.
+            --
+            queryParent p = Concurrently $ void $ do
+                logg Debug $ taskMsg k $ "getBlockHeaderInternal.getPrerequisteHeader (parent) for " <> sshow h <> ": " <> sshow p
+                ChainValue _ ph <- getBlockHeaderInternal headerStore payloadStore candidateHeaderCas candidatePayloadCas priority maybeOrigin' p
+                validateInductiveM ph header
```

Comment:
Copied `queryPrerequesiteHeader` into `queryAdjacentParent` and `queryParent`. The former is unchanged the latter got the `validateInductive ph header` check on the result.